src/storage.rs

```diff
@@ -73,42 +82,64 @@ where
     /// # Examples
 
     // @TODO specify more useful error type
-    pub fn write(&mut self, record: Record<K>) -> Result<(), ()> {
-        self.initialized.ok_or(())?;
-        if self.is_active_blob_full(record.size()) {
-            let new_active = Box::new(Default::default());
+    pub fn write(&mut self, record: &mut Record) -> Result<WriteFuture> {
+        if self.is_active_blob_full(record.full_len())? {
+            let next = self.next_blob_name()?;
+            let new_active = Blob::open_new(next).map_err(Error::BlobError)?.boxed();
             // @TODO process unwrap explicitly
-            let mut old_active = self.active_blob.replace(new_active).unwrap();
-            old_active.flush()?;
+            let old_active = self.active_blob.replace(new_active).unwrap();
             self.blobs.push(*old_active);
         }
         // @TODO process unwrap explicitly
-        self.active_blob.as_mut().unwrap().write(record)
+        self.active_blob
+            .as_mut()
+            .unwrap()
+            .write(record)
+            .map_err(Error::BlobError)
+    }
+
+    fn next_blob_name(&self) -> Result<blob::FileName> {
+        let active_blob_id = self.active_blob.as_ref().map(|blob| blob.id());
+        let blobs_max_id = self.blobs.iter().max_by_key(|blob| blob.id()).map(Blob::id);
```

Comment:
Suboptimal approach. It is better to keep BLOBs in order and simply add 1 to `active_blob_id` to calculate `next_id`

Reply:
reworked